kernel/src/common/regs/README.md

```diff
@@ -0,0 +1,266 @@
+    // or u32.
+    u8,
+
+    // Each subsequent parameter is a register abbreviation, its descriptive
+    // name, and its associated bitfields.
+    // The descriptive name defines this 'group' of bitfields. Only registers
+    // defined as ReadWrite<_, Control> can use these bitfields.
+    CR Control [
+        // Bitfields are defined as:
+        // name (Mask(mask), shift) [ /* optional named values */ ]
+
+        // This is a two-bit field which includes bits 4 and 5
+        RANGE (Mask(0b11), 4) [
+            // Each of these defines a name for a value that the bitfield can be
+            // written with or matched against. Note that this set is not exclusive--
+            // the field can still be written with arbitrary constants.
+            VeryHigh = 0,
+            High = 1,
+            Low = 2
+        ],
+
+        // A common case is single-bit bitfields, which usually just mean
+        // 'enable' or 'disable' something. This syntax is a shorthand, defining
+        // EN as a field over bit 3, and INT as a field over bit 2.
+        EN 3 [],
+        INT 2 []
+    ],
+
+    // Without the explanatory comments like above, bitfield definition is quite compact:
+    // Status register
+    S Status [
+        TXCOMPLETE 0 [],
+        TXINTERRUPT 1 [],
+        RXCOMPLETE 2 [],
+        RXINTERRUPT 3 [],
+        MODE (Mask(0b11), 4) [
+            FullDuplex = 0,
+            HalfDuplex = 1,
+            Loopback = 2,
+            Disabled = 3
+        ],
+        ERRORCOUNT (Mask(0b11), 6) []
+    ]
+]
+```
+
+## Register Interface Summary
+
+There are three types provided by the register interface: `ReadOnly`, `WriteOnly`, and `ReadWrite`. They provide the following functions:
+
+```rust
+ReadOnly<T: IntLike, R: RegisterLongName = ()>
+.get() -> T                                 // Get the raw register value
+.read(field: Field<T, R>) -> T              // Read the value of the given field
+.is_set(field: Field<T, R>) -> bool         // Check if a one-bit field is set
+.matches(value: FieldValue<T, R>) -> bool   // Check if one or more fields match a value
+
+WriteOnly<T: IntLike, R: RegisterLongName = ()>
+.set(value: T)                              // Set the raw register value
+.write(value: FieldValue<T, R>)             // Write the value of one or more fields,
+                                            //  overwriting other fields to zero
+
+ReadWrite<T: IntLike, R: RegisterLongName = ()>
+.get() -> T                                 // Get the raw register value
+.set(value: T)                              // Set the raw register value
+.read(field: Field<T, R>) -> T              // Read the value of the given field
+.write(value: FieldValue<T, R>)             // Write the value of one or more fields,
+                                            //  overwriting other fields to zero
+.modify(value: FieldValue<T, R>)            // Write the value of one or more fields,
```

Comment:
I'm worried that the difference between `write` and `modify` are non-obvious. I expected write to have the `modify` behavior. Do we actually need the current `write` behavior? I imagine if you wanted to set everything else to zero, you could use `set`.

Reply:
The setting everything to zero is actually really just a side-effect of the more fundamental difference between `write` and `modify`. The real difference is that `write` uses a single store operation, whereas `modify` is a single load followed by a single store (eg a read-modify-write). In general, this distinction is important.

Reply:
I agree that it makes sense with regards to single store and load/store. However, I don't think the naming makes it very clear what the result is.

Reply:
I had that issue for a second as well when I first read this, but then I couldn't come up with any better names, and decided this was actually a pretty good interface.

Reply:
My proposal would be:

 1. `write` for WriteOnly registers to be the same as it is now.
 2. `write` should be totally removed for ReadWrite registers, as I think it is confusing.
 3. `write_and_clear_others` would do a single store to ReadWrite registers. (Name is a suggestion, still thinking of possible better names)
 4. `modify` should stay the same for ReadWrite registers.

Reply:
I think if we change the `write` name, we should change it for both WriteOnly and ReadWrite registers. Having `write` on one and `write_and_clear_others` on the other implies that they have different behavior, which is even more confusing.

Reply:
I thought about changing `modify()` to `update()` and `write()` to `overwrite()`, but after thinking more I'm not sure it adds any clarity. If there's a word out there that captures that the entire register is changing, and not just the passed in fields, and we can come up with it, that would be great, but otherwise I think just learning what `write()` does the best option.

Reply:
So the concern on my side, and lets see if I sell you on this @bradjc, is that if I call `write` and pass in a specific field that I'm writing to, I do not expect any side effects on the rest of the register. And in my mind, writing all the other fields in the register to zero is certainly a side effect.

To me, the difference is that I'm specifying a field to write to, and therefore expect it to write only that field.

Reply:
My other proposal here would be for `write` to "always be side effect free". For write-only registers, it would just keep doing what it's doing. For read/write registers, it would do a load and a store in order to only change the bits specified.

Then you could remove `modify` altogether. If you still needed to write and zero out other bits as a single operation, you could use the `set` function to do so.

Reply:
I'm completely buying what you are selling, but I think we have to support the case where you can set a register with the pretty named values and do it in only one store. The cost is that you have to read the docs closely, and I'm ok with that.

Reply:
I don't think you should conflate `set` into the operations here. It takes a manually-crafted `value`, not `Field`s, so you have none of the type safety to construct what you're writing. `set` is not a replacement for anything.

What about `overwrite` or maybe `rewrite` in place of the current `write`, but otherwise the same interface?